tests/load/gradual_orchestrator.go

```diff
@@ -0,0 +1,272 @@
+func (o *GradualOrchestrator[T, U]) run(ctx context.Context) bool {
+	var (
+		prevConfirmed        = o.tracker.GetObservedConfirmed()
+		prevTime             = time.Now()
+		currTargetTPS        = new(atomic.Uint64)
+		attempts      uint64 = 1
+		// true if the orchestrator has reached the max TPS target
+		achievedTargetTPS bool
+	)
+
+	currTargetTPS.Store(o.config.MinTPS)
+
+	issuerGroup, issuerCtx := errgroup.WithContext(ctx)
+	o.issuerGroup = issuerGroup
+
+	o.issueTxs(issuerCtx, currTargetTPS)
+
+	for {
+		// wait for the sustained time to pass or for the context to be cancelled
+		select {
+		case <-time.After(o.config.SustainedTime):
+		case <-issuerCtx.Done(): // the parent context was cancelled or an issuer errored
+		}
+
+		if issuerCtx.Err() != nil {
+			break // Case 1
+		}
+
+		currConfirmed := o.tracker.GetObservedConfirmed()
+		currTime := time.Now()
+
+		tps := computeTPS(prevConfirmed, currConfirmed, currTime.Sub(prevTime))
+		o.setMaxObservedTPS(tps)
+
+		// if max TPS target has been reached and we don't terminate, then continue here
+		// so we do not keep increasing the target TPS
+		if achievedTargetTPS && !o.config.Terminate {
+			o.log.Info(
+				"current network state",
+				zap.Uint64("current TPS", tps),
+				zap.Uint64("max observed TPS", o.maxObservedTPS.Load()),
+			)
+			continue
+		}
+
+		if tps >= currTargetTPS.Load() {
+			if currTargetTPS.Load() >= o.config.MaxTPS {
+				achievedTargetTPS = true
+				o.log.Info(
+					"max TPS target reached",
+					zap.Uint64("max TPS target", currTargetTPS.Load()),
+					zap.Uint64("average TPS", tps),
+				)
+				if o.config.Terminate {
+					o.log.Info("terminating orchestrator")
+					break // Case 2
+				} else {
+					o.log.Info("orchestrator will now continue running at max TPS")
+					continue
+				}
+			}
+			o.log.Info(
+				"increasing TPS",
+				zap.Uint64("previous target TPS", currTargetTPS.Load()),
+				zap.Uint64("average TPS", tps),
+				zap.Uint64("new target TPS", currTargetTPS.Load()+o.config.Step),
+			)
+			currTargetTPS.Add(o.config.Step)
+			attempts = 1
+		} else {
+			if attempts >= o.config.MaxAttempts {
+				o.log.Info(
+					"max attempts reached",
+					zap.Uint64("attempted target TPS", currTargetTPS.Load()),
+					zap.Uint64("number of attempts", attempts),
+				)
+				break // Case 3
+			}
+			o.log.Info(
+				"failed to reach target TPS, retrying",
+				zap.Uint64("current target TPS", currTargetTPS.Load()),
+				zap.Uint64("average TPS", tps),
+				zap.Uint64("attempt number", attempts),
+			)
+			attempts++
+		}
+
+		prevConfirmed = currConfirmed
+		prevTime = currTime
+	}
+
+	return achievedTargetTPS
+}
+
+// GetObservedIssued returns the max TPS the orchestrator observed
+func (o *GradualOrchestrator[T, U]) GetMaxObservedTPS() uint64 {
+	return o.maxObservedTPS.Load()
+}
+
+// start a goroutine to each issuer to continuously send transactions
+// if an issuer errors, all other issuers will stop as well.
```

Comment:
nit adding a full stop otherwise it sounds "send transactions if an issuer errors" 😄 

```suggestion
// start a goroutine to each issuer to continuously send transactions.
// if an issuer errors, all other issuers will stop as well.
```